tests/fuzz/fuzz_exceptions.py

```diff
@@ -0,0 +1,548 @@
+    if type(node) != list:
+        return [node]
+    return node
+
+
+def indent(code):
+    return '\n'.join(['  ' + line for line in code.splitlines() if line])
+
+
+class Cursor:
+    '''
+    A cursor over an array, allowing gradual consumption of it. If we run out, we
+    return simple values.
+    '''
+    def __init__(self, array):
+        self.array = arrayify(array)
+        self.pos = 0
+
+    def get(self):
+        if self.pos >= len(self.array):
+            return 0
+        self.pos += 1
+        return self.array[self.pos - 1]
+
+    def get_num(self):
+        return numify(self.get())
+
+    def get_array(self):
+        return arrayify(self.get())
+
+    def remaining(self):
+        return max(0, len(self.array) - self.pos)
+
+    def has_more(self):
+        return self.remaining() > 0
+
+    def get_int(self):
+        return int(1000 * self.get_num())
+
+
+def pick(options, value, param):
+    '''
+    Given a list of options (weight, func), and a value in [0, 1) to help pick from
+    them, pick one, and call it with the parameter.
+    '''
+    # Scale the value by the total weight.
+    assert 0 <= value < 1, value
+    total = 0
+    for weight, func in options:
+        total += weight
+    value *= total
+
+    for weight, func in options:
+        if value < weight:
+            return func(param)
+        value -= weight
+
+    raise Exception('inconceivable')
+
+
+class CppTranslator:
+    '''
+    Translates random structured data into a random C++ program that uses C++
+    exceptions.
+    '''
+
+    PREAMBLE = '''\
+#include <stdio.h> // avoid iostream C++ code, just test libc++abi, not libc++
+#include <stdint.h>
+
+extern void refuel();
+extern void checkRecursion();
+extern bool getBoolean();
+'''
+
+    SUPPORT = '''\
+#include <stdio.h>
+#include <stdlib.h>
+
+const int INIIAL_FUEL = 100;
+
+static int fuel = INIIAL_FUEL;
+
+void refuel() {
+  fuel = INIIAL_FUEL;
+}
+
+void checkRecursion() {
+  if (fuel == 0) {
+    puts("out of fuel");
+    abort();
+  }
+  fuel--;
+}
+
+// TODO random data
+static bool boolean = true;
+
+bool getBoolean() {
+  // If we are done, exit all loops etc.
+  if (fuel == 0) {
+    return false;
+  }
+  fuel--;
+  boolean = !boolean;
+  return boolean;
+}
+'''
+
+    def __init__(self, data):
+        self.toplevel = Cursor(data)
+        self.logging_index = 0
+        self.try_nesting = 0
+        self.loop_nesting = 0
+
+        # The output is a list of strings which will be concatenated when
+        # writing.
+        self.output = [self.PREAMBLE]
+        self.make_functions()
+
+    '''
+    Outputs the main file and the support file on the side. Support code is not
+    in the main file so that the optimizer cannot see it all.
+    '''
+    def write(self, main, support):
+        with open(main, 'w') as f:
+            f.write('\n'.join(self.output))
+        with open(support, 'w') as f:
+            f.write(self.SUPPORT)
+
+    def make_structs(self):
```

Comment:
It looks like the structs aren't actually used for anything?

Reply:
Yes, they are for future use.